Firestore/core/src/firebase/firestore/remote/write_stream.mm

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/firebase/firestore/remote/write_stream.h"
+
+#include "Firestore/core/src/firebase/firestore/util/hard_assert.h"
+#include "Firestore/core/src/firebase/firestore/util/log.h"
+#include "Firestore/core/src/firebase/firestore/util/status.h"
+
+#import "Firestore/Protos/objc/google/firestore/v1beta1/Firestore.pbobjc.h"
+
+namespace firebase {
+namespace firestore {
+namespace remote {
+
+using auth::CredentialsProvider;
+using util::AsyncQueue;
+using util::TimerId;
+using util::Status;
+
+WriteStream::WriteStream(AsyncQueue* async_queue,
+                         CredentialsProvider* credentials_provider,
+                         FSTSerializerBeta* serializer,
+                         Datastore* datastore,
+                         id<FSTWriteStreamDelegate> delegate)
+    : Stream{async_queue, credentials_provider, datastore,
+             TimerId::WriteStreamConnectionBackoff, TimerId::WriteStreamIdle},
+      serializer_bridge_{serializer},
+      delegate_bridge_{delegate} {
+}
+
+void WriteStream::SetLastStreamToken(NSData* token) {
+  serializer_bridge_.SetLastStreamToken(token);
+}
+
+NSData* WriteStream::GetLastStreamToken() const {
+  return serializer_bridge_.GetLastStreamToken();
+}
+
+void WriteStream::WriteHandshake() {
+  EnsureOnQueue();
+  HARD_ASSERT(IsOpen(), "Writing handshake requires an opened stream");
+  HARD_ASSERT(!IsHandshakeComplete(), "Handshake already completed");
+
+  GCFSWriteRequest* request = serializer_bridge_.CreateHandshake();
+  LOG_DEBUG("%s initial request: %s", GetDebugDescription(),
+            serializer_bridge_.Describe(request));
+  Write(serializer_bridge_.ToByteBuffer(request));
+
+  // TODO(dimond): Support stream resumption. We intentionally do not set the
+  // stream token on the handshake, ignoring any stream token we might have.
+}
+
+void WriteStream::WriteMutations(NSArray<FSTMutation*>* mutations) {
+  EnsureOnQueue();
+  HARD_ASSERT(IsOpen(), "Handshake already completed");
```

Comment:
The error message looks wrong. How about "Writing mutations requires an opened stream" to mirror the error message in WriteHandshake()?

Reply:
Done, thanks for spotting this.